### PR TITLE
Ensure test context throws exception if further assertions made after context completed/failed

### LINKED ISSUE #106: Test context completion state sanity checks
#### Describe the feature

`VertxTestContext` throws an appropriate exception to signal an invalid state if a new checkpoint or handler (`succeeding`/`failing`/`verify` etc.) is created when the test context is already completed (when the `releaseLatch` inside the context  equals 0). 

#### Use cases

I'm creating this feature request because I recently hit an issue in the Strimzi codebase where tests were passing despite deliberate failures being introduced into their `context.verify()` blocks. 

This occurred because a single checkpoint had been created and then flagged as completed before the async Vert.x code under test was actually exercised, and when the last remaining checkpoint is flagged, the context is marked as completed. Further checkpoints were subsequently created, but whether or not they were flagged was irrelevant, as the release latch was already at zero. Likewise, if any failures had been caught, they would only be registered with the test context if it isn't marked as completed.

As all the code under test was running in a separate thread,  the unit test method didn't block, and when `VertxExtension` used `awaitsCompletion` in an `afterEach` context, to wait for the async code under test, no blocking occurred, because the test context was already marked as completed, so the `context.verify()` block never managed to run before all threads exited, and as such never managed to throw an `AssertionError` that would cause the test to fail.

While the usage of the checkpoint I encountered was not how checkpoints are intended to be used, I would like add checks to catch this edge case to prevent tests passing when they should fail.

#### Contribution

I am very willing to submit a PR for this, I just wanted to check with the contributors that I a) understand the logic flow of the `VertxExtension` correctly and b) that I am not missing a use case that needs to create handlers and checkpoints after a test context is complete.

#### Reproduction case

https://gist.github.com/LiamClarkeNZ/945c37ab094ae88ca47b279e9b3e7448

```
package io.strimzi.operator;

import io.vertx.core.Future;
import io.vertx.junit5.Checkpoint;
import io.vertx.junit5.VertxExtension;
import io.vertx.junit5.VertxTestContext;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

import static org.junit.jupiter.api.Assertions.assertEquals;

@ExtendWith(VertxExtension.class)
public class ReproductionTest {

    @Test
    void reproduceCheckpointIssues(VertxTestContext context) {

        Checkpoint prematurelyFlagged = context.checkpoint();
        prematurelyFlagged.flag();
        
        Checkpoint neverCompletedCheckpointShouldTimeoutTestButDoesNot = context.checkpoint();
        
        Future.succeededFuture("Example").onComplete(context.succeeding(example -> context.verify(() -> {
            assertEquals("Break this assertion", example);
            throw new RuntimeException("Die");
        })));
    }
}
```        
**Output:**

```
Process finished with exit code 0
```

**Example desired output:**

```
java.lang.IllegalStateException: Cannot create new checkpoint on already completed test context

<stacktrace>

Process finished with exit code 255
```

_Or_

```
java.lang.IllegalStateException: Cannot create new succeeding() handler on already completed test context

<stacktrace>

Process finished with exit code 255
```

### PR DESCRIPTION
**Pull request by LiamClarkeNZ**
Motivation:

Currently it's possible to write tests where the test context is accidentally completed, causing further assertions later in the code to be ignored. This could cause tests to pass when they should fail. Please see #106 for more details.

Conformance:

I've just signed the ECA, but the check isn't able to match me up yet.

### CODE REVIEW

**Comment by LiamClarkeNZ** [comment]
@jponge - very sorry for this very delayed PR! 

I've submitted this as a draft PR, as I'd really appreciate your thoughts on my approach, I'm uncertain if I've gone overboard (e.g., I had to change some existing unit tests because they created a `succeeding` or `failing` block after the context had completed or failed). 

So very keen for your thoughts :)

File: src/main/java/io/vertx/junit5/VertxTestContext.java [hunk 1]
```diff
@@ -113,6 +114,11 @@ public synchronized void completeNow() {
    */
   public synchronized void failNow(Throwable t) {
     Objects.requireNonNull(t, "The exception cannot be null");
+
+    // Unsure on this, but don't want to prevent the overuse of a strict checkpoint's flag() failing the test with its own exception
+    if (!(t instanceof IllegalStateException)) {
+      assertContextNotCompletedOrFailed();
+    }
```
Comment (LiamClarkeNZ) [inline src/main/java/io/vertx/junit5/VertxTestContext.java:121]: @jponge - I added this clause as calling flag() more than once on a strict checkpoint fails the context with an IllegalStateException, and I didn't want the new ISE I added to override that one. But it doesn't feel like a great approach.

File: src/main/java/io/vertx/junit5/VertxTestContext.java [hunk 2]
```diff
@@ -392,4 +409,18 @@ public boolean awaitCompletion(long timeout, TimeUnit unit) throws InterruptedEx
...
+  /**
+   * If the context has already been completed or failed, then throw an exception to prevent code under test from
+   * not being properly tested
+   * @throws IllegalStateException if the context is completed or failed
+   */
+  private void assertContextNotCompletedOrFailed() throws IllegalStateException {
+    if (failed() || completed()) {
+      String errorMessage = "Vert.x test context has already succeeded or failed. This can occur when a checkpoint is flagged " +
+        "earlier in your test, or completeNow or failNow were called. If you're using multiple checkpoints, ensure that you " +
+        "create them upfront before flagging them";
```
Comment (LiamClarkeNZ) [inline src/main/java/io/vertx/junit5/VertxTestContext.java:422]: @jponge - is the error message okay?

File: src/main/java/io/vertx/junit5/VertxTestContext.java (hunk 1 above)
Comment (jponge) [reply src/main/java/io/vertx/junit5/VertxTestContext.java:121]: The problem is that you could have an `IllegalStateException` for some other reason than a strict checkpoint failing

File: src/main/java/io/vertx/junit5/VertxTestContext.java (hunk 2 above)
Comment (jponge) [reply src/main/java/io/vertx/junit5/VertxTestContext.java:422]: That's descriptive, great!

File: src/test/java/io/vertx/junit5/VertxTestContextTest.java [hunk 3]
```diff
@@ -335,10 +337,7 @@ void check_future_completion_failure() throws InterruptedException {
     VertxTestContext context = new VertxTestContext();
     context
       .assertComplete(Future.succeededFuture("bla"))
-      .compose(s -> context.assertComplete(Future.failedFuture(new IllegalStateException(s + "bla"))))
```
Comment (jponge) [inline src/test/java/io/vertx/junit5/VertxTestContextTest.java:338]: Why did you remove these lines?

Comment (LiamClarkeNZ) [reply src/test/java/io/vertx/junit5/VertxTestContextTest.java:338]: Yeah, so this might be where my changes went a bit too far, in `assertComplete` I added `   assertContextNotCompletedOrFailed()`, so the second assertComplete throws. Probably makes more sense to remove that assertion from `assertComplete` and `assertFail`.

File: src/main/java/io/vertx/junit5/VertxTestContext.java (hunk 1 above)
Comment (LiamClarkeNZ) [reply src/main/java/io/vertx/junit5/VertxTestContext.java:121]: Good point. What if I subclassed IllegalStateException for `checkpoint.flag()` to throw, so I could special case that precisely? It would also preserve the API if someone is catching ISE around a call to `flag` for some reason.

Comment (LiamClarkeNZ) [reply src/main/java/io/vertx/junit5/VertxTestContext.java]: Hi @jponge I've pushed another version up, when you have time to look at it :)

**Comment by julianladisch** [comment]
This implementation is too excessive because it fails tests that have been correctly succeeding for a long time.

Example:

```
  // method under test
  static void divisible(long l, Handler<Long> divisibleBy2, Handler<Long> divisibleBy5) {
    if (l % 2 == 0) {
      divisibleBy2.handle(l);
    }
    if (l % 5 == 0) {
      divisibleBy5.handle(l);
    }
  }

  static Stream<Arguments> testData() {
    return Stream.of(
        Arguments.of(2),
        Arguments.of(4),
        Arguments.of(5),
        Arguments.of(6),
        Arguments.of(8),
        Arguments.of(10));
  }

  @ParameterizedTest
  @MethodSource("testData")
  void is_divisible_by_2_or_5(long l, VertxTestContext vtc) {
    divisible(l, x -> vtc.completeNow(), x -> vtc.completeNow());
  }
```

The reproduction case posted in the description of #106 has been fixed with PR #114 released as Vert.x 4.2.6.
I don't see a need for further changes.
Please provide a new reproduction case that shows that the problem still exists.

**Comment by jponge** [comment]
Closing due to a lack of activity.
